### ISSUE TITLE
clang::Expr::ClassifyImpl asserts on isPRValue()

### ISSUE THREAD

**Issue by vgvassilev** [llvm/llvm-project#53958]
[Sema::FixOverloadedFunctionReference](https://github.com/llvm/llvm-project/blob/896a543e72fd3ab044e64d1140a37c7f4876fc71/clang/lib/Sema/SemaOverload.cpp#L15076) where we seem to create a lvalue DeclRefExpr [here](https://github.com/llvm/llvm-project/blob/896a543e72fd3ab044e64d1140a37c7f4876fc71/clang/lib/Sema/SemaOverload.cpp#L15184) but then make a possibly `UnaryOperator` [here](https://github.com/llvm/llvm-project/blob/896a543e72fd3ab044e64d1140a37c7f4876fc71/clang/lib/Sema/SemaOverload.cpp#L15160). I suspect that's not expected, if it was then we need to rework the `clang::Sema::BuildUnaryOp` which eventually calls `clang::Expr::ClassifyImpl` which has `assert(isPRValue());`. The example program is pretty trivial: https://godbolt.org/z/ffbcneqzM

[The C++ ](https://eel.is/c++draft/expr.prim.id#qual-5 says that the name of a non-static member function is an lvalue if it's the operand of an &, so one can argue that `FixOverloadedFunctionReference` is "right" here, and the "regular" codepath is wrong.

However, the fix is going to be quite hard as `ExprClassification` assumes that the category does not depend on the context of use. We should probably ignore what the standards says and "fix" `FixOverloadedFunctionReference` to return an rvalue and add a `FIXME` note.

cc: @zygoloid

**Comment by jmciver** [comment]
I would like to work on this issue. I have started looking into it.

**Comment by vgvassilev** [comment]
There is a patch already here https://reviews.llvm.org/D122103

**Comment by fhahn** [comment]
@vgvassilev do you know if the issue has been fixed already? I can't access the patch you linked because it has some kind of access restriction set.

**Comment by vgvassilev** [comment]
@fhahn, I don't think it was fixed I saw the patch, however, then it was made private...

**Comment by fhahn** [comment]
Oh that's a shame!

**Comment by shafik** [comment]
The reproducer you linked: https://godbolt.org/z/ffbcneqzM

Does not crash, is that the correct reproducer?

**Comment by vgvassilev** [comment]
@shafik, yes, it is the correct reproducer. We are not expecting a crash but inconsistency between the interfaces which breaks the AST invariants. You can see what kind of workarounds we need to introduce [here](https://github.com/vgvassilev/clad/pull/484/commits/8a462140ad3e6369b1bede2d67f51f4723557b23).
